pkg/rule/rule.go

```diff
@@ -0,0 +1,171 @@
+package thanosrule
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/rules"
+
+	"github.com/prometheus/tsdb"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const tmpRuleDir = ".tmp-rules"
+
+type Group struct {
+	*rules.Group
+
+	PartialResponseStrategy storepb.PartialResponseStrategy
+}
+
+type AlertingRule struct {
+	*rules.AlertingRule
+
+	PartialResponseStrategy storepb.PartialResponseStrategy
+}
+
+type RuleGroups struct {
+	Groups []RuleGroup `yaml:"groups"`
+}
+
+type RuleGroup struct {
+	rulefmt.RuleGroup
+
+	PartialResponseStrategy storepb.PartialResponseStrategy `yaml:"partial_response_strategy"`
+}
+
+type Managers map[storepb.PartialResponseStrategy]*rules.Manager
+
+func (m Managers) RuleGroups() []Group {
+	var res []Group
+	for s, r := range m {
+		for _, group := range r.RuleGroups() {
+			res = append(res, Group{Group: group, PartialResponseStrategy: s})
+		}
+	}
+	return res
+}
+
+func (m Managers) AlertingRules() []AlertingRule {
+	var res []AlertingRule
+	for s, r := range m {
+		for _, r := range r.AlertingRules() {
+			res = append(res, AlertingRule{AlertingRule: r, PartialResponseStrategy: s})
+		}
+	}
+	return res
+}
+
+func (r *RuleGroup) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	rs := struct {
+		String string `yaml:"partial_response_strategy"`
+	}{}
+
+	errMsg := fmt.Sprintf("failed to unmarshal 'partial_response_strategy'. Possible values are %s", strings.Join(storepb.PartialResponseStrategyValues, ","))
+	if err := unmarshal(&rs); err != nil {
+		return errors.Wrapf(err, errMsg)
+	}
+
+	rg := rulefmt.RuleGroup{}
+	if err := unmarshal(&rg); err != nil {
+		return errors.Wrapf(err, errMsg)
+	}
+
+	p, ok := storepb.PartialResponseStrategy_value[strings.ToUpper(rs.String)]
+	if !ok {
+		if rs.String != "" {
+			return errors.Errorf("%s. Got: %s", errMsg, rs.String)
+		}
+
+		// NOTE: For Rule default is abort as this is recommended for alerting.
+		p = storepb.PartialResponseStrategy_value[storepb.PartialResponseStrategy_ABORT.String()]
```

Comment:
@GiedriusS in terms of backward compatibility we break it here. By default you will have `abort` strategy. (it was "enabled"/"warning" previously only). Added item to CHANGELOG as well.